DataFormats/Provenance/src/CompactHash.cc

```diff
@@ -0,0 +1,71 @@
+#include "DataFormats/Provenance/interface/CompactHash.h"
+#include "FWCore/Utilities/interface/Algorithms.h"
+#include "FWCore/Utilities/interface/Digest.h"
+#include "FWCore/Utilities/interface/EDMException.h"
+
+#include <functional>
+#include <cassert>
+
+namespace {
+  std::array<unsigned char, 16> convert(std::string const& v) {
+    assert(v.size() == 16);
+    std::array<unsigned char, 16> retValue;
+    std::copy(v.begin(), v.end(), retValue.begin());
+    return retValue;
+  }
+}  // namespace
+namespace edm {
+  namespace detail {
+    // This string is the 16-byte, non-printable version.
+    std::array<unsigned char, 16> const& InvalidCompactHash() {
+      static std::array<unsigned char, 16> const invalid = convert(cms::MD5Result().compactForm());
+      return invalid;
+    }
+  }  // namespace detail
+
+  namespace compact_hash_detail {
+    size_t smallHash_(value_type const& hash) {
+      //NOTE: In future we could try to xor the first 8bytes into the second 8bytes of the string to make the hash
+      std::hash<std::string_view> h;
+      return h(std::string_view(reinterpret_cast<const char*>(hash.data()), hash.size()));
+    }
+
+    std::array<unsigned char, 16> fromHex_(std::string_view v) {
+      cms::MD5Result temp;
+      temp.fromHexifiedString(std::string(v));
```

Comment:
Since this function will unconditionally create `std::string`, I think, in principle, it would be better to have the argument type as `std::string const&`. In practice that would only move the `std::string` creation to the `CompactHash` constructor calling `fromHex_()`. On the other hand, that would require `CompactHash.h` to `#include <string>`. I'm not really sure if the change would be worth it.

Reply:
It looks like it would be trivial to change `cms::MD5Result::fromHexifiedString` to take a `std::string_view`.

Reply:
done